File: src/main/java/com/gorogoro/auth/global/exception/GlobalExceptionHandler.java
package com.gorogoro.auth.global.exception;

import com.gorogoro.auth.global.exception.code.GlobalErrorCode;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.List;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    /**
     * [Exception] Valid 유효성 검사 실패
     */
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleValidationException(MethodArgumentNotValidException ex) {
        BindingResult bindingResult = ex.getBindingResult();

        List<ErrorResponse.FieldError> fieldErrors = bindingResult.getFieldErrors().stream()
                .map(error -> new ErrorResponse.FieldError(
                        error.getField(),
                        error.getDefaultMessage(),
                        error.getRejectedValue()
                ))
                .toList();

        ErrorCode errorCode = GlobalErrorCode.INVALID_REQUEST;
        log.warn("ValidationException occurred: [{}] {}", errorCode.getCode(), errorCode.getMessage());

        return ResponseEntity
                .status(errorCode.getHttpStatus())
                .body(ErrorResponse.of(errorCode.getCode(), errorCode.getMessage(), fieldErrors));
    }

    /**
     * [Exception] BaseException (커스텀 비즈니스 예외)
     */
    @ExceptionHandler(BaseException.class)
    public ResponseEntity<ErrorResponse> handleBaseException(BaseException ex) {
        ErrorCode errorCode = ex.getErrorCode();
        log.warn("BaseException occurred: [{}] {}", errorCode.getCode(), ex.getMessage());

        return ResponseEntity
                .status(errorCode.getHttpStatus())
                .body(ErrorResponse.of(errorCode.getCode(), errorCode.getMessage()));
    }

    /**
     * [Exception] 예상치 못한 예외
     */
    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleGeneralException(Exception ex) {
        ErrorCode internalError = GlobalErrorCode.INTERNAL_SERVER_ERROR;
        log.warn("Unexpected exception occurred", ex);

        return ResponseEntity
                .status(internalError.getHttpStatus())
                .body(ErrorResponse.of(internalError.getCode(), internalError.getMessage()));
    }
}

File: src/main/kotlin/com/gorogoro/auth/global/exception/GlobalExceptionHandler.kt
package com.gorogoro.auth.global.exception

import com.gorogoro.auth.global.exception.dto.ErrorResponse
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(BusinessException::class)
    fun handleBusinessException(e: BusinessException): ResponseEntity<ErrorResponse> {

        val response = ErrorResponse(
            message = e.message,
            code = e.CustomErrorCode
        )

        return ResponseEntity.status(e.httpStatus).body(response)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValidException(e: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
        val errorField = e.bindingResult.fieldErrors.firstOrNull()?.field

        val errorCode = ValidationErrorCode.findByField(errorField) ?: ValidationErrorCode.INVALID_INPUT_IN_COMMAND

        val response = ErrorResponse(
            message = errorCode.message,
            code = errorCode.errorCode
        )

        return ResponseEntity
            .status(errorCode.httpStatus)
            .body(response)
    }


    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<ErrorResponse> {
        val error = ErrorCode.GLOBAL_ERROR_UNEXPECTED
        val response = ErrorResponse(
            message = error.message,
            code = error.errorCode
        )

        return ResponseEntity.status(error.status).body(response)
    }
}
